### PR TITLE
Add DEBUG-level logging to the "http" state storage backend

### PR DESCRIPTION
**Pull request by haiderbari**
Resolves # feature request 2003 https://github.com/opentofu/opentofu/issues/2003

## Target Release

1.9.0, backport the change

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

**Comment by haiderbari** [comment]
example of the CLI logs
2024-10-17T12:30:57.837+0100 [DEBUG] Executing HTTP remote state request for: "unlock"
2024-10-17T12:30:57.837+0100 [DEBUG] UNLOCK http://localhost:9000/JamesGosling/unlock
Releasing state lock. This may take a few moments...
2024-10-17T12:30:58.457+0100 [DEBUG] HTTP remote state request for "unlock" returned status code: 200
2024-10-17T12:30:58.457+0100 [DEBUG] HTTP response headers: {"Content-Type":"application/json"}

2024-10-17T12:30:59.231+0100 [DEBUG] Executing HTTP remote state request for: "get state"
2024-10-17T12:30:59.231+0100 [DEBUG] HTTP remote state request for "get state" returned status code: 403
2024-10-17T12:30:59.231+0100 [DEBUG] GET STATE, Forbidden: "{"error":"Forbidden"}"

File: internal/backend/remote-state/http/client.go
Comment (haiderbari) [reply internal/backend/remote-state/http/client.go:52]: The issue stated using debug logs since the user would want to know why they are getting non 200 response and if you look at the other backend clients especially oss it uses DEBUG.

Comment (haiderbari) [reply internal/backend/remote-state/http/client.go:52]: Do you mean only the logs within httpRequest method or all the DEBUG logs

Comment (haiderbari) [reply internal/backend/remote-state/http/client.go:52]: Thanks I understand I will update the log at line 52 to a trace log

File: internal/backend/remote-state/http/helper.go
Comment (ollevche) [inline internal/backend/remote-state/http/helper.go:16]: nit: `parseResponseBody` and `parseHeaders` should only be used for logging purposes and I would suggest reflecting this in their naming. Something like `prepareResponseBodyForLog` / `parseResponseBodyForLog`

File: internal/backend/remote-state/http/helper.go
Comment (ollevche) [inline internal/backend/remote-state/http/helper.go:35]: HTTP headers are case-insensitive so we must normalize the values before comparison. Also, headers in the blacklist map should be already normalized (e.g. in lowercase).

File: internal/backend/remote-state/http/helper_test.go
Comment (ollevche) [inline internal/backend/remote-state/http/helper_test.go:83]: Let's extend the test with lower/upper cased headers as well.

**Review by ollevche — CHANGES_REQUESTED** [review]
Thanks for the contribution! Left a few minor suggestions and headers-normalizing comment (with a blocking change due to security risks).

**Comment by haiderbari** [comment]
Hi @apparentlymart can you please review this PR, I am currently working on issue 2027. I recently had one PR merge which caused a merge conflict. Would like to get this moved on so I can resolve any more merge conflicts once the branch is merged. Thank you.

**Comment by ollevche** [comment]
Hey @haiderbari! We discussed this PR with the core team yesterday and it is good to be merged with two approvals.

**Pull request merged**
